File: cs/days/day01.cs
using System.Collections.Immutable;

namespace Shunty.AoC.Days;

// https://adventofcode.com/2024/day/1 - Historian Hysteria

public class Day01 : AocDaySolver
{
    public int DayNumber => 1;

    public async Task Solve()
    {
        var lines = (await File.ReadAllLinesAsync(AocUtils.FindInputFile(DayNumber)))
            .Select(x => {
                var parts = x.Split("  ");
                return (int.Parse(parts[0]), int.Parse(parts[^1]));
            })
            .ToImmutableList();
        var left1 = lines.Select(x => x.Item1).OrderBy(x => x).ToImmutableList();
        var right1 = lines.Select(x => x.Item2).OrderBy(x => x).ToImmutableList();

        var part1 = left1.Zip(right1, (x, y) => Math.Abs(x - y)).Sum();

        this.ShowDayHeader();
        this.ShowDayResult(1, part1);

        // Part 2
        // "Calculate a total similarity score by adding up each number in the left
        // list after multiplying it by the number of times that number appears in the right list."
        var left2 = left1.GroupBy(x => x).ToImmutableDictionary(x => x.Key, x => x.Count());
        var right2 = right1.GroupBy(x => x).ToImmutableDictionary(x => x.Key, x => x.Count());

        var part2 = left2.Sum(x => right2.ContainsKey(x.Key) ? x.Key * x.Value * right2[x.Key] : 0);
        this.ShowDayResult(2, part2);
    }
}

File: go/days/day01.go
package days

// https://adventofcode.com/2024/day/1 - Historian Hysteria

import (
	aoc "aoc2024/aocutils"
	"slices"
)

func Day01() {
	lines, err := aoc.GetDayInputLines(1)
	aoc.CheckError(err)

	l := make([]int, len(lines))
	r := make([]int, len(lines))
	lunique := make(map[int]bool)
	rmap := make(map[int]int)
	for i, line := range lines {
		nums, err := aoc.StringToInts(line, "   ")
		aoc.CheckError(err)
		l[i] = nums[0]
		r[i] = nums[1]

		lunique[nums[0]] = true
		rmap[nums[1]] = rmap[nums[1]] + 1
	}

	slices.Sort(l)
	slices.Sort(r)
	part1 := 0
	for i := 0; i < len(l); i++ {
		d := aoc.Abs(l[i] - r[i])
		part1 += d
	}

	aoc.DayHeader(1)
	aoc.PrintResult(1, part1)

	part2 := 0
	for v := range lunique {
		part2 += rmap[v] * v
	}
	aoc.PrintResult(2, part2)
}
